DEVDOCS-2076: Adding JavaScript (#615)
bigcommerce/dev-docs: docs/stencil-docs/javascript-and-event-hooks/adding-javascript.md

### Minification

Beyond reducing the number of HTTP calls required to fetch all the required JavaScript modules, we can reduce the size of the individual JavaScript modules through minification. JavaScript minification removes white space and comments, shortens variable and function names, removes dead code, and more. The goal, in all cases, is to reduce the amount of bandwidth necessary to transmit the JavaScript module to the browser.

## Development options

When you add JavaScript to a theme, use one of the following techniques, so that Stencil will automatically bundle and minify your modules.

* **Using npm:** Add third-party JavaScript modules to your theme with npm where possible.
* **Place modules in assets/js/:** For a JavaScript module that is not distributed via npm, add this module to your theme by creating a subdirectory within assets/js/ that contains your module.
* **Theme-specific JavaScript modules:** Stencil themes include their own custom JavaScript modules for most page types. You can alter these page-type–specific modules by editing the files in assets/js/theme/*.js.

These techniques are outlined in the following sections.

> ## Using npm (Node Package Manager)

Reviewer: ```suggestion
## Using npm
```

> Many third-party JavaScript components are distributed with npm (Node Package Manager). When you use the npm command-line utility to add a JavaScript component to your theme, Stencil will automatically bundle and minify the component. To enable this bundling and minification, run each module's `npm install` command from the root directory of your theme.

Reviewer: ```suggestion
Many third-party JavaScript components are distributed with npm. When you use the npm command-line utility to add a JavaScript component to your theme, Stencil will automatically bundle and minify the component. To enable this bundling and minification, run each module's `npm install` command from the root directory of your theme.
```

### Taking over from npm

> npm facilitates managing third-party JavaScript components by placing each JavaScript component in the correct directories. However, as a developer, you will still need to edit your theme files to wire up the JavaScript component to expose it on your storefront. You will find several examples of this on the following pages.

Reviewer: ```suggestion
The npm package manager facilitates the management of third-party JavaScript components by placing each JavaScript component in the correct directories. However, as a developer, you will still need to edit your theme files to wire up the JavaScript component to expose it on your storefront. You will find several examples of this on the following pages.
```

        }
```

## Mapping custom templates to JavaScript modules

If you add [custom page templates](/stencil-docs/storefront-customization/custom-templates/) to your theme, you can edit the same `assets/js/app.js` file to map each custom template to an appropriate JavaScript module.

In the following example, we will map a custom JavaScript file, `assets/js/themes/custom.js` to a custom product page file `templates/pages/custom/product/customProd.html`.

```js
import PageManager from "./page-manager";

export default class Custom extends PageManager {
    onReady(){
        alert("Hello world!");
    }
}
```

This is a basic module that creates a class called `Custom` which extends the PageManager class.

```html
{{~inject 'template' template}}
<h2>Hello World!</h2>
<body>
Some custom content!
<body>
<script>window.__webpack_public_path__ = "{{cdn 'assets/dist/'}}";</script>
<script src="{{cdn 'assets/dist/theme-bundle.main.js'}}"></script>
<script>window.stencilBootstrap("{{page_type}}", {{jsContext}}).load();</script>
```

> In order to successully map your custom module to a custom template file, that file must do the following:

Reviewer: Do the following need to be done in this specific order? If so, this should be an ordered list.

Author: Nah, its just a general list of requirements.
